### PR TITLE
Copy parent in copy_similar for Symmetric/Hermitian

### PR DESCRIPTION
**Pull request by jishnub**
This should be equivalent to copying the `Symmetric`/`Hermitian` matrix, as only one triangular half is used. However, this reduces latency when the parent is a `Matrix` (and possibly others too, although I've not checked this).

```julia
julia> using LinearAlgebra

julia> A = rand(2,2); H = Hermitian(A);

julia> @time eigen(H);
  0.342207 seconds (440.61 k allocations: 22.926 MiB, 99.95% compilation time) # nightly
  0.260913 seconds (326.62 k allocations: 16.926 MiB, 99.93% compilation time) # This PR
```
Copying the parent is also marginally faster, although this doesn't really matter in `eigen`:
```julia
julia> A = rand(1000,1000); H = Hermitian(A);

julia> @btime LinearAlgebra.copy_similar($H, Float64);
  1.839 ms (3 allocations: 7.63 MiB) # nightly
  1.760 ms (3 allocations: 7.63 MiB) # This PR
```

### CODE REVIEW

**Comment by fredrikekre** [comment]
> 99.95% compilation time

What is the timing without compilation? I had a quick look and the only difference is that Symmetric goes through https://github.com/JuliaLang/julia/blob/5949e0bdfca4981ee10588aafbfcf0a4c7f13fec/stdlib/LinearAlgebra/src/symmetric.jl#L317 which shouldn't cost anything compared to what you implement here, or?

**Comment by jishnub** [comment]
This PR is primarily about faster compilation. The difference mainly arises from the fact that `copyto!` is specialized for `Array`s to essentially be reduced to a foreigncall to `jl_genericmemory_copyto`. On the other hand, `copyto!` for a `Hermitian` uses the fallback `copyto_unaliased!` path, which takes longer to compile.
```julia
julia> using LinearAlgebra

julia> A = rand(1000,1000); H = Hermitian(A); B = similar(A);

julia> @time copyto!(B, A);
  0.024590 seconds (38.30 k allocations: 1.943 MiB, 81.51% compilation time)
```
vs
```julia
julia> @time copyto!(B, H);
  0.122573 seconds (178.02 k allocations: 9.301 MiB, 13.93% gc time, 95.64% compilation time)
```
Even leaving the compilation-time aside, the `copyto!` is faster for `Array`-to-`Array` copies:
```julia
julia> @btime copyto!($B, $A);
  475.371 μs (0 allocations: 0 bytes)

julia> @btime copyto!($B, $H);
  1.737 ms (0 allocations: 0 bytes)
```
However, in this specific context, this difference may not matter much, if at all.

**Comment by dkarrasch** [comment]
What if we specialized `copy_similar` to `HermOrSym`? We could copy then only half of the data from the parent (i.e., without `getindex` passing through the `HermOrSym` branches)?

**Comment by jishnub** [comment]
The docstring of `copy_similar` currently says that the type of the output corresponds to `similar(A, T, size(A))` in general (although I'm unsure why the three-term `similar` is specifically necessary). If we relax the output type, we should be able to use `copytrito!` and wrap the result in a `Symmetric`/`Hermitian`, which would be faster in general.

File: stdlib/LinearAlgebra/src/symmetriceigen.jl [hunk 1]
```diff
@@ -1,8 +1,8 @@
 # This file is a part of Julia. License is MIT: https://julialang.org/license
 
 # preserve HermOrSym wrapper
-eigencopy_oftype(A::Hermitian, S) = Hermitian(copy_similar(A, S), sym_uplo(A.uplo))
-eigencopy_oftype(A::Symmetric, S) = Symmetric(copy_similar(A, S), sym_uplo(A.uplo))
+eigencopy_oftype(A::Hermitian, S) = Hermitian(copy_similar(parent(A), S), sym_uplo(A.uplo))
+eigencopy_oftype(A::Symmetric, S) = Symmetric(copy_similar(parent(A), S), sym_uplo(A.uplo))
```
Comment (dkarrasch) [inline stdlib/LinearAlgebra/src/symmetriceigen.jl:5]: You mean like this?

```suggestion
eigencopy_oftype(A::Hermitian, S) = Hermitian(copytrito!(similar(parent(A), S, size(A)), A, A.uplo), sym_uplo(A.uplo))
eigencopy_oftype(A::Symmetric, S) = Symmetric(copytrito!(similar(parent(A), S, size(A)), A, A.uplo), sym_uplo(A.uplo))
```

That indeed reduces the runtime by a factor of 2, as expected, and the compilation time by one third.

**Comment by dkarrasch** [comment]
BTW, should we make use of `copytrito!` in `copyto!(::Hermitian, ::Hermitian)`? We could then try something else: define `eigencopy_oftype(A::Hermitian, S) = copymutable_oftype(A, S)`. If I made no mistake, this reduces the runtime by another factor of 3, but increases compile time slightly relative to the status quo.

**Comment by dkarrasch** [comment]
I went ahead and opened #54476, since it's orthogonal to this PR, but allows for more options.

**Comment by jishnub** [comment]
I don't think we should call `copyto!` here, but we should call `copytrito!` directly on the parents (if not `copyto!`). This is because we'll unnecessarily compile the mismatched-`uplo` branch in `copyto!`. In a way, we're inlining the matched-`uplo` branch of `copyto!`.

**Comment by jishnub** [comment]
I've timed the `copytrito!` and the `copyto_similar` implementations, and they seem roughly comparable:

With
```julia
eigencopy_oftype(A::Hermitian, S) = Hermitian(copytrito!(similar(parent(A), S, size(A)), A.data, A.uplo), sym_uplo(A.uplo))
```
we obtain
```julia
julia> using LinearAlgebra

julia> A = rand(200,200); H = Hermitian(A);

julia> @time eigen(H);
  0.299632 seconds (318.16 k allocations: 17.390 MiB, 97.93% compilation time)

julia> @btime eigen!($H);
  5.333 ms (20 allocations: 698.94 KiB)
```
whereas, with the `copy_similar` implementation as in this PR, we obtain
```julia
julia> @time eigen(H);
  0.274765 seconds (326.63 k allocations: 17.926 MiB, 98.01% compilation time)

julia> @btime eigen!($H);
  5.505 ms (20 allocations: 698.94 KiB)
```
Perhaps we should go with `copytrito!` then, as it'll lower run-times for smaller matrices.

File: stdlib/LinearAlgebra/src/symmetriceigen.jl (hunk 1 above)
Comment (jishnub) [reply stdlib/LinearAlgebra/src/symmetriceigen.jl]: I've implemented this now

Comment (dkarrasch) [reply stdlib/LinearAlgebra/src/symmetriceigen.jl]: The reason I found a performance gap is what you found: it should call `copytrito!` on `A.data`, not on `A`!

**Review by dkarrasch — APPROVED** [review]

**Comment by dkarrasch** [comment]
Hm, failures look real. Something about reference to undefined objects in `Matrix{BigFloat}`...

**Comment by jishnub** [comment]
Yeah, that's one potential catch with using `copytrito!`, where one half of the matrix is left untouched (and potentially uninitialized). In this case, the issue seems to be that `adjoint!` and `transpose!` don't support `undef` values in a matrix. Perhaps it's prudent to use `copyto!` for now.

**Comment by dkarrasch** [comment]
I don't have time right now to work it out, but as an option, can we catch the potentially bad cases in `adjoint!` and `transpose!` by types? Like overload these two for triangular matrices or something? One would need to see the full signature in those failing cases, which is impossible from the stacktrace because almsot all the function calls are inlined. I'll try to have a look later.

**Comment by jishnub** [comment]
The specific failing test should be fixed by https://github.com/JuliaLang/julia/pull/54491

**Comment by jishnub** [comment]
The failing `BunchKaufman` test should be fixed by https://github.com/JuliaLang/julia/pull/54509

**Comment by jishnub** [comment]
The failing `hessenberg` test should be fixed by https://github.com/JuliaLang/julia/pull/54518 (I think)

**Comment by dkarrasch** [comment]
Awesome, how many further improvements this has triggered through failing tests. :rofl:

**Pull request merged**